pkg/controller/lambda/function/setup.go

```diff
@@ -0,0 +1,516 @@
+	}
+
+	return nil
+}
+
+func preObserve(_ context.Context, cr *svcapitypes.Function, obj *svcsdk.GetFunctionInput) error {
+	obj.FunctionName = aws.String(meta.GetExternalName(cr))
+	return nil
+}
+
+func postObserve(_ context.Context, cr *svcapitypes.Function, resp *svcsdk.GetFunctionOutput, obs managed.ExternalObservation, err error) (managed.ExternalObservation, error) {
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+	switch aws.StringValue(resp.Configuration.State) {
+	case string(svcapitypes.State_Active):
+		cr.SetConditions(xpv1.Available())
+	case string(svcapitypes.State_Pending):
+		cr.SetConditions(xpv1.Creating())
+	case string(svcapitypes.State_Failed):
+		cr.SetConditions(xpv1.Unavailable())
+	case string(svcapitypes.State_Inactive):
+		cr.SetConditions(xpv1.Unavailable())
+	}
+	return obs, nil
+}
+
+func preDelete(_ context.Context, cr *svcapitypes.Function, obj *svcsdk.DeleteFunctionInput) (bool, error) {
+	obj.FunctionName = aws.String(meta.GetExternalName(cr))
+	return false, nil
+}
+
+// nolint:gocyclo
+func isUpToDate(cr *svcapitypes.Function, obj *svcsdk.GetFunctionOutput) (bool, error) {
+
+	// Compare CODE
+	// GetFunctionOutput returns
+	// Code *FunctionCodeLocation `type:"structure"`
+	// which does not map to
+	// Code *FunctionCode `type:"structure" required:"true"`
+	// which is used when creating the function.
+	// We can't currently properly implement a comparison
+
+	// Compare CONFIGURATION
+	if aws.StringValue(cr.Spec.ForProvider.Description) != aws.StringValue(obj.Configuration.Description) {
+		return false, nil
+	}
+
+	if !isUpToDateEnvironment(cr, obj) {
+		return false, nil
+	}
+
+	// Connection settings for an Amazon EFS file system.
+	if !isUpToDateFileSystemConfigs(cr, obj) {
+		return false, nil
+	}
+
+	if aws.StringValue(cr.Spec.ForProvider.Handler) != aws.StringValue(obj.Configuration.Handler) {
+		return false, nil
+	}
+
+	if aws.StringValue(cr.Spec.ForProvider.KMSKeyARN) != aws.StringValue(obj.Configuration.KMSKeyArn) {
+		return false, nil
+	}
+
+	// The function's layers (https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html).
+	// The generator is creating layers with type of
+	// 	Layers []*string `json:"layers,omitempty"`
+	// Instead of []*svcsdk.Layer
+	// We can't properly implement a comparison until that is fixed
+
+	// set default
+	if aws.Int64Value(cr.Spec.ForProvider.MemorySize) != aws.Int64Value(obj.Configuration.MemorySize) {
+		return false, nil
+	}
+
+	if aws.StringValue(cr.Spec.ForProvider.Role) != aws.StringValue(obj.Configuration.Role) {
+		return false, nil
+	}
+
+	if aws.StringValue(cr.Spec.ForProvider.Runtime) != aws.StringValue(obj.Configuration.Runtime) {
+		return false, nil
+	}
+
+	if aws.Int64Value(cr.Spec.ForProvider.Timeout) != aws.Int64Value(obj.Configuration.Timeout) {
+		return false, nil
+	}
+
+	if aws.StringValue(cr.Spec.ForProvider.TracingConfig.Mode) != aws.StringValue(obj.Configuration.TracingConfig.Mode) {
```

Comment:
`cr.Spec.ForProvider.TracingConfig` or `obj.Configuration.TracingConfig.Mode` might be `nil` and this could result in panic. We need to account for that case, too.

Reply:
added a comment in the code - it can never be nil as it will be set via LateInitialize with the default value from aws, if not set